cou/utils/openstack.py

```diff
@@ -13,118 +13,125 @@
-        else:
-            return OPENSTACK_CODENAMES[vers]
+"""Lookup utils to have the latest compatible Openstack codename based on workload version."""
 
+import csv
+import logging
+from collections import OrderedDict, defaultdict
+from dataclasses import dataclass
+from typing import Any, Union
 
-# Codename and package versions
-def get_swift_codename(version):
-    """Determine OpenStack codename that corresponds to swift version.
+from packaging.version import Version
 
-    :param version: Version of Swift
-    :type version: string
-    :returns: Codename for swift
-    :rtype: string
-    """
-    return _get_special_codename(version, SWIFT_CODENAMES)
 
+@dataclass(frozen=True)
+class Versions:
+    min_version: Version
+    max_version: Version
+    minor: bool = False
+    micro: bool = False
 
-def get_ovn_codename(version):
-    """Determine OpenStack codename that corresponds to OVN version.
 
-    :param version: Version of OVN
-    :type version: string
-    :returns: Codename for OVN
-    :rtype: string
-    """
-    return _get_special_codename(version, OVN_CODENAMES)
+def generate_openstack_lookup() -> OrderedDict:
+    """Generate an OpenStack lookup dictionary based on the version of the components.
+
+    The dictionary is generated thru an static csv file that should be update regularly
+    to include new Openstack releases and update the min and max versions of the services.
+    The csv table is made from the release page [0] charm delivery [1], cmadison and rmadison.
 
+    In the csv file it's possible to see that there are two columns named "minor" and "micro".
+    Set to true if the service does not get major version updates between OpenStack releases.
+    E.g:
+    ussuri: 10.1.3 focal 11.2.3. In this case the identifier is the "major" (10 and 11).
+    ussuri: 17.1.6 focal: 17.2.3. In this case the identifier is the "minor" (17.1 and 17.2).
+    ussuri: 13.1.0 focal 13.1.1. In this case the identifier is the "micro" (13.1.0 and 13.1.1).
 
-def _get_special_codename(version, codenames):
-    found = [k for k, v in six.iteritems(codenames) if version in v]
-    return found[0]
+    charm designate-bind workload_version tracks the version of the deb package bind9.
+
+    [0] https://releases.openstack.org/
+    [1] https://docs.openstack.org/charm-guide/latest/project/charm-delivery.html
+
+    :return: Ordered dictionary containing the version and the possible Openstack release.
+    :rtype: OrderedDict
+    """
+    charms_types = {
+        "ceph": ["ceph-mon", "ceph-fs", "ceph-radosgw", "ceph-osd"],
+        "swift": ["swift-proxy", "swift-storage"],
+        "nova": ["nova-cloud-controller", "nova-compute"],
+        "ovn": ["ovn-dedicated-chassis", "ovn-central"],
+        "neutron": ["neutron-api", "neutron-gateway"],
+        "manila": ["manila-ganesha"],
+        "horizon": ["openstack-dashboard"],
+    }
+    with open("cou/utils/openstack_lookup.csv") as csv_file:
```

Comment:
The path should be defined as constant.

Reply:
parameter of `__init__` method would be nicer(still constant).

Reply:
it could be default value as constant, but you are right it can be configurable. However with __init__ this object will be generate anytime when it's called, so I think using `__new__` would be better here.